upstream/upstream.go

```diff
@@ -654,11 +658,103 @@ func (u *Upstream) shouldSkip(req *common.NormalizedRequest) (reason error, skip
+	bn, ebn := req.EvmBlockNumber()
+	if ebn != nil || bn <= 0 {
+		return fmt.Errorf("failed to get valid request block number: %w", ebn), false
+	}
+
+	if u.config.Evm.NodeType == common.EvmNodeTypeFull {
+		if req.Network() == nil || req.Network().EvmStatePollerOf(u.Config().Id) == nil {
+			return fmt.Errorf("network or state poller is nil"), false
+		}
+		lb := req.Network().EvmStatePollerOf(u.Config().Id).LatestBlock()
+
+		if bn < lb-u.config.Evm.MaxAvailableRecentBlocks {
+			return common.NewErrEndpointMissingData(fmt.Errorf("block number %d is out of range (must be >= %d)", bn, lb-u.config.Evm.MaxAvailableRecentBlocks)), true
+		}
+
+	}
 
 	return nil, false
 }
 
+func (u *Upstream) detectNodeType() (common.EvmNodeType, error) {
+	// Step 1: Get the latest block number using the state poller
+	req := &common.NormalizedRequest{}
+	if req.Network() == nil || req.Network().EvmStatePollerOf(u.Config().Id) == nil {
+		return "", common.NewErrUpstreamClientInitialization(fmt.Errorf("network or state poller is nil"), u.config.Id)
+	}
+	lb := req.Network().EvmStatePollerOf(u.Config().Id).LatestBlock()
+
+	// Block heights to check
+	blockHeights := []int64{
+		1,
+		lb / 2,
+		lb - 1024,
+		lb - 128,
+		lb - 32,
+	}
+
+	// Track responses for each block height
+	responses := make(map[int64]bool)
+
+	for _, blockHeight := range blockHeights {
+		// Normalize the block height to hex format
+		hexBlockHeight, err := common.NormalizeHex(fmt.Sprintf("%x", blockHeight))
+
+		if err != nil {
+			return "", err
+		}
+
+		balanceReq := common.NewNormalizedRequest([]byte(fmt.Sprintf(
+			`{"jsonrpc":"2.0","id": %d,"method":"eth_getBlockByNumber","params":["%s", false]}`,
+			util.RandomID(),
+			hexBlockHeight,
+		)))
+
+		resp, errForward := u.Forward(context.Background(), balanceReq)
+		if errForward != nil {
+			return "", errForward
+		}
+
+		if errForward == nil {
+			rpcResp, err := resp.JsonRpcResponse()
+			if err != nil {
+				return "", err
+			}
+			blockRef, _ := rpcResp.PeekStringByPath("hash")
+
+			// Check if the response contains a valid block
+			if rpcResp != nil && rpcResp.Result != nil && blockRef != "" {
+				responses[blockHeight] = true
+			} else {
+				responses[blockHeight] = false
+			}
+		} else if common.EvmIsMissingDataError(errForward) {
+			// Error due to missing historical state, continue checking
+			responses[blockHeight] = false
+		} else {
+			// Log the error for this specific block height
+			u.Logger.Debug().Err(errForward).Int64("blockHeight", blockHeight).Msg("Error getting balance")
+			responses[blockHeight] = false
+		}
+	}
+
+	// Determine node type based on responses
+	isArchiveNode := responses[1] && responses[lb/2] && responses[lb-1024]
+	isFullNode := !isArchiveNode && (responses[lb-128] || responses[lb-32])
+
+	if isArchiveNode {
+		return "archive", nil
+	} else if isFullNode {
+		return "full", nil
+	}
+
+	// If none of the checks succeeded, it's likely a light node
+	return "light", nil
```

Comment:
Better to use the const instead of hardcoded strings. It is common.EvmNodeTypeFull or something like that iirc